blog post Extend Falco inputs by creating a Plugin: Distribute the plugin (#988)
falcosecurity/falco-website: content/en/blog/extend-falco-inputs-with-a-plugin-distribute/index.md

> See other articles:
> * [Extend Falco inputs by creating a Plugin: the basics]({{< ref "/blog/extend-falco-inputs-with-a-plugin-the-basics" >}})
> * [Extend Falco inputs by creating a Plugin: Register the plugin]({{< ref "/blog/extend-falco-inputs-with-a-plugin-register" >}})

----

## Introduction

In the previous posts (see [the basics]({{< ref "/blog/extend-falco-inputs-with-a-plugin-the-basics" >}}) and [register]({{< ref "/blog/extend-falco-inputs-with-a-plugin-register" >}})), we covered all the basics to develop a plugin and register it. In this article, we'll focus on the steps to build the OCI artifacts and distribute them on Github Packages.

{{% pageinfo color="info" %}}
To get more familiar with the OCI artifacts, you can read our blog posts about [falcoctl]({{< ref "falcoctl-install-manage-rules-plugins" >}}) and [GitOps for rules]({{< ref "gitops-your-falco-rules" >}})
{{% /pageinfo %}}

In the next sections we'll describe how to:
* set up a Github Actions workflow to:
  * create a release with `goreleaser` when a tag is pushed
  * build the OCI artifacts of the plugin and its rules
* create the `index.yaml` used by `falcoctol`

> ## Table of contents

Author: I'm removing the block. Thanks

- [Goreleaser](#goreleaser)
- [Makefile](#makefile)
- [Github Actions Workflow](#github-actions-workflow)
  - [Headers](#headers)
  - [The jobs](#the-jobs)
    - [Publish the OCI artifacts](#publish-the-oci-artifacts)
    - [Create the release](#create-the-release)
  - [Final result](#final-result)
- [index.yaml file](#indexyaml-file)
- [Create our first version](#create-our-first-version)
- [Installation of your plugin and rules](#installation-of-your-plugin-and-rules)
- [Conclusion](#conclusion)

## Requirements

This tutorial is based on a Github repo, with the possibility to run workflows in Github Actions and store OCI artifacts in Github Packages. If you use a different system or even just private repositories, you'll need to adapt the examples to your context.

{{% pageinfo color="warning" %}}
To make it works, you must have followed the code organization proposed in [the basics]({{< ref "/blog/extend-falco-inputs-with-a-plugin-the-basics#code-organization" >}}) post.
{{% /pageinfo %}}

## Goreleaser

`Goreleaser` is a pretty famous tool to build and create releases for projects on Github or else. We'll use it build and archive the binaries at each release.

> At the root of your repo, create a `.goreleaser.yaml` file with this content:

Author: Noted. Thanks.

checksum:
  name_template: "checksums.txt"
```

with `PLUGIN_NAME` the name of your plugin.

## Makefile

`Makefiles` are a convenient way to script actions, we'll use to pass all requirement flags to create the finale `.so` library files, used by the plugin framework of `Falco`:

```makefile
SHELL=/bin/bash -o pipefail
GO ?= go

NAME := {PLUGIN_NAME}
OUTPUT := lib$(NAME).so

ifeq ($(DEBUG), 1)
    GODEBUGFLAGS= GODEBUG=cgocheck=2
else
    GODEBUGFLAGS= GODEBUG=cgocheck=0
endif

all: build

clean:
	@rm -f lib$(NAME).so

build: clean
	@$(GODEBUGFLAGS) $(GO) build -buildmode=c-shared -buildvcs=false -o $(OUTPUT) ./plugin
```

with `PLUGIN_NAME` the name of your plugin.

You can test it by running `make build` and see the `lib{PLUGIN_NAME}.so` appearing at the root of your folder.

## Github Actions Workflow

The first step is to create the folders `.github/workflows` and the file describinng our workflow `release.yaml`.

### Headers

Each workflow starts with, at least, a `name` and a `on`:

```yaml
name: Release Plugins

on:
  push:
    tags:
      - '*'

env:
  OCI_REGISTRY: ghcr.io
  PLUGIN_NAME: {PLUGIN_NAME}

permissions:
  contents: write
  packages: write
```

> The workflow will be triggered each time a tag `x.y.z` is created.

Author: It's a requirement for falcoctl to follow a branch (2.x or 2.1.x). I edited the regexp.

{{% pageinfo color="warning" %}}
The permissions are required to allow Github Actions to read the content of your repo, create the release and push the artifacts into Github Packages.
{{% /pageinfo %}}

### The jobs

Once we have set the "headers" of the workflow file, it's time to set the actions that will be run. We'll split them into 2:
* Publish the OCI artifacts
* Create the release

#### Publish the OCI artifacts

To publish the artifacts, we'll use [`falcoctl`](https://github.com/falcosecurity/falcoctl), the official CLI to manage the Falco artifacts.

The actions to publish the artifacts will be, in this order:

* Get the falcoctl sources
* Prepare the Go env
* Build falcoctl (to be sure to have always the last version as the tool moves a lot)
* Get the plugin sources
* Build the plugin .so
* Get the repo name in lower case
* Push the artifacts with all their tags:
  * push the plugin
  * push the rules with a dependency to the plugin

```yaml
jobs:
  publish-oci-artifacts:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Falcoctl Repo
        uses: actions/checkout@v3
        with:
          repository: falcosecurity/falcoctl

>           ref: main

Author: I changed for `0.5.0`

          path: tools/falcoctl
      - name: Setup Golang
        uses: actions/setup-go@v4
        with:
          go-version: '^1.20'
          cache-dependency-path: tools/falcoctl/go.sum
      - name: Build falcoctl
        run: make
        working-directory: tools/falcoctl
      - name: Checkout
        uses: actions/checkout@v3
        with:
          path: plugin
      - name: Build the plugin
        run: make build
        working-directory: plugin
      - id: StringRepoName
        uses: ASzc/change-string-case-action@v5
        with:
          string: ${{ github.repository }}
      - name: Upload OCI artifacts to GitHub packages
        run: |
              MAJOR=$(echo ${{ github.ref_name }} | cut -f1 -d".")
              MINOR=$(echo ${{ github.ref_name }} | cut -f1,2 -d".")
              DIR=$(pwd)

              cd plugin/
              $DIR/tools/falcoctl/falcoctl registry push \
              ${{ env.OCI_REGISTRY }}/${{ steps.StringRepoName.outputs.lowercase }}/plugin/${{ env.PLUGIN_NAME }}:${{ github.ref_name }} \
              --config /dev/null \
              --type plugin \
              --version "${{ github.ref_name }}" \

>               --tag latest --tag $MAJOR --tag $MINOR \

Reviewer: Does that mean that you want the latest from that major and that minor (e.g. 0.5.x)?

              --platform linux/amd64 \
              --requires plugin_api_version:2.0.0 \
              --depends-on ${{ env.PLUGIN_NAME }}-rules:${{ github.ref_name }} \
              --name ${{ env.PLUGIN_NAME }} \
              lib${{ env.PLUGIN_NAME }}.so
    
              cd rules/
              $DIR/tools/falcoctl/falcoctl registry push \
              ${{ env.OCI_REGISTRY }}/${{ steps.StringRepoName.outputs.lowercase }}/ruleset/${{ env.PLUGIN_NAME }}:${{ github.ref_name }} \
              --config /dev/null \
              --type rulesfile \
              --version "${{ github.ref_name }}" \
              --tag latest --tag $MAJOR --tag $MINOR \
              --depends-on ${{ env.PLUGIN_NAME }}:${{ github.ref_name }} \
              --name ${{ env.PLUGIN_NAME }}-rules \
              ${{ env.PLUGIN_NAME }}_rules.yaml
        env:
          FALCOCTL_REGISTRY_AUTH_BASIC: ${{ env.OCI_REGISTRY }},${{ github.repository_owner }},${{ secrets.GITHUB_TOKEN }}
```

#### Create the release

> At the same time we publish new artifacts, we'll also create a release with an automatic changelog. It's done by [`Goreleaser`](https://goreleaser.com/). This section is out of context of this blog post, it's a classic with Go programs, just take care to have a correct `.goreleaser` as explained [here]({{< ref "/blog/extend-falco-inputs-with-a-plugin-distribute##goreleaser" >}}).

Reviewer: This sentence is long and weird. I can't follow what you are trying to say. It is out of scope, but you are adding it. It is also a simple config, which would be easy to explain. I think that a quick explanation paragraph would be easier than this entire text to say that it is out of scope.

Reviewer 2: I'd rather say something like 'out of scope', not out of context.

What about:
`GoReleaser` can automatically generate a _Changelog_ at the same time we publish the new artifacts. This step isn't imperative to generate the OCI artifacts but it's a good practice among Go developers. To achieve that, make sure to have a correct `.goreleaser` file as explained here.

Author: I changed for Vic's version. Thanks.

              --config /dev/null \
              --type rulesfile \
              --version "${{ github.ref_name }}" \
              --tag latest --tag $MAJOR --tag $MINOR \
              --depends-on ${{ env.PLUGIN_NAME }}:${{ github.ref_name }} \
              --name ${{ env.PLUGIN_NAME }}-rules \
              ${{ env.PLUGIN_NAME }}_rules.yaml
        env:
          FALCOCTL_REGISTRY_AUTH_BASIC: ${{ env.OCI_REGISTRY }},${{ github.repository_owner }},${{ secrets.GITHUB_TOKEN }}

  release:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          fetch-depth: 0
      - name: Setup Golang
        uses: actions/setup-go@v3
        with:
          go-version: '1.19'
      - name: Run GoReleaser
        uses: goreleaser/goreleaser-action@v4
        with:
          version: latest
          args: release --clean --timeout 120m
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          LDFLAGS: "-buildmode=c-shared"
          GOPATH: /home/runner/go
```

Replace `PLUGIN_NAME` with the name of your plugin.

## index.yaml file

> This file is used by `falcoctl` to know where to download the plugins and rules. We'll create our own file to allow anyone to use it locally.

Reviewer: Not sure if you cover this in the other posts (I didn't read them). This sentence gives me a general idea but is too vague. It would be great to have a better understanding of how it works with `falcoctl`. For example, Why does creating our own file allows anyone to use it locally? Is there one `index.yaml` file per plugin? If so, "plugins" in the plural form doesn't help my understanding.

Author: I added a link to https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index to explain

```yaml
- name: {PLUGIN_NAME}
  type: plugin
  registry: ghcr.io
  repository: {OWNER_NAME}/{REPO_NAME}/plugin/{PLUGIN_NAME}

>   description: Reads events from Docker Engine

Reviewer: Suddenly it is not a generic plugin anymore. Is this on purpose?

Author: Wrong copy/paste

  home: https://github.com/{OWNER_NAME}/{PLUGIN_NAME}
  keywords:
    - {PLUGIN_NAME}
  license: Apache-2.0
  maintainers:
    - email: {OWNER_EMAIL}
      name: {OWNER_REAL_NAME}
  sources:
    - https://github.com/{OWNER_NAME}/{PLUGIN_NAME}
- name: {PLUGIN_NAME}-rules
  type: rulesfile
  registry: ghcr.io
  repository: {OWNER_NAME}/{REPO_NAME}/ruleset/docker

>   description: Rules for the Docker plugin

Author: Wrong copy/paste too. Fixed.

  home: https://github.com/{OWNER_NAME}/{REPO_NAME}/tree/main/rules
  keywords:
    - {PLUGIN_NAME}
  license: Apache-2.0
  maintainers:
    - email: {OWNER_EMAIL}
      name: {OWNER_REAL_NAME}
  sources:
    - https://github.com/{OWNER_NAME}/{REPO_NAME}/tree/main/rules/{PLUGIN_NAME}_rules.yaml
```

With:
* `PLUGIN_NAME`: the name of you plugin
* `OWNER_NAME`: your nickname on Github
* `REPO_NAME`: the name of your repo for your plugin on Github
* `OWNER_EMAIL`: an email for contact
* `OWNER_REAL_NAME`: your real name or not

> The `keywords` are used for the search feature of `falcoctl`, try add to add relevant terms for your plugin.

Reviewer 2: I'd rather say:
`falcoctl` uses the `keywords` field to perform a search among your plugins. Leverage this functionality by adding relevant terms to your plugin.

You should now have a structure of repo like this:

```shell.
├── .github
│   └── workflows
│       └── release.yaml
├── .gitignore
├── go.mod
├── .goreleaser.yml
├── go.sum
├── index.yaml
├── LICENSE
├── Makefile
├── README.md
├── pkg
│   └── {PLUGIN_NAME}.go
├── plugin
│   └── main.go
└── rules
    └── {PLUGIN_NAME}_rules.yaml
```

There is 2 ways to expose the `index.yaml`:
* exposing the raw file: `https://raw.githubusercontent.com/{OWNER_NAME}/{REPO_NAME}/main/index.yaml`

> * exposing the file through Github Page: `https://{OWNER_NAME}.github.io/{REPO_NAME}/index.yaml` (you need to [enable the Pages](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-from-a-branch) of course)

Reviewer: ```suggestion
* exposing the file through Github Page: `https://{OWNER_NAME}.github.io/{REPO_NAME}/index.yaml` (make sure to [enable the Pages](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-from-a-branch))
```

## Create our first version

Everything is now ready to publish a first version of our plugin.

In you `main` branch, run:

```shell
git tag 0.1.0 -m "0.1.0" && git push origin 0.1.0
```

> Few seconds after, your workflow should be started and you will have your first published version with associated artifacts (for example, with the `docker` plugin):

Author: Good point. I started a new sentence. What are you calling the "rest of the code"?

> Few seconds after, your workflow should be started and you will have your first published version with associated artifacts (for example, with the `docker` plugin):

Author: The tagging is manual, so I don't see any issue. You can't even tag in the Github UI, it must be done through git cli (locally or in a CI workflow).

![](images/workflow.png)
![](images/packages.png)

## Installation of your plugin and rules

The process is now exactly the same than the one described [here]({{< ref "/blog/falcoctl-install-manage-rules-plugins" >}}), except we'll use you specific `index.yaml` to register a new index:

```shell
sudo falcoctl index add {PLUGIN_NAME} https://{OWNER_NAME}.github.io/{REPO_NAME}/index.yaml
sudo falcoctl artifact install {PLUGIN_NAME}
sudo falcoctl artifact install {PLUGIN_NAME}-rules
```

For the `docker` plugin, for example:

```shell
❯ sudo falcoctl index add docker  http://issif.github.io/docker-plugin/index.yaml

> ❯ sudo falcoctl artifact search docker

Author: Yep, to show we can now search for a plugin not in the official index.
